File: src/plugins/metrics.js
import http from 'node:http';
import fp from 'fastify-plugin';
import client from 'prom-client';

/**
 * Prometheus metrics plugin (search-api).
 *
 * Exposes RED metrics (Rate / Errors / Duration) for every HTTP route plus a
 * few search-domain counters, and serves them on a DEDICATED internal port
 * (never on the public app port).
 *
 * PM2 cluster note: prom-client's AggregatorRegistry must run in the cluster
 * primary, which PM2 owns — so cross-worker aggregation is not possible here.
 * Instead each worker exposes its own metrics endpoint on
 *   METRICS_PORT_BASE + NODE_APP_INSTANCE
 * and Prometheus scrapes every worker (see monitoring/prometheus/prometheus.yml).
 * Aggregation across workers is done at query time with `sum by (job)`.
 */

const SERVICE_NAME = process.env.SERVICE_NAME || 'search-api';
const METRICS_PORT_BASE = parseInt(process.env.METRICS_PORT_BASE || '9101', 10);
const METRICS_HOST = process.env.METRICS_HOST || '0.0.0.0';

function resolveMetricsPort() {
    // PM2 sets NODE_APP_INSTANCE to the 0-based worker index in cluster mode.
    const instance = parseInt(process.env.NODE_APP_INSTANCE || '0', 10);
    return METRICS_PORT_BASE + (Number.isNaN(instance) ? 0 : instance);
}

async function metricsPlugin(fastify) {
    const register = new client.Registry();
    register.setDefaultLabels({ service: SERVICE_NAME });
    client.collectDefaultMetrics({ register });

    const httpHistogram = new client.Histogram({
        name: 'http_request_duration_seconds',
        help: 'HTTP request duration in seconds',
        labelNames: ['method', 'route', 'status_code'],
        buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10],
        registers: [register]
    });

    // --- Search domain metrics ---
    const searchRequests = new client.Counter({
        name: 'search_requests_total',
        help: 'Total search requests handled, by type',
        labelNames: ['type'],
        registers: [register]
    });
    const searchCacheHits = new client.Counter({
        name: 'search_cache_hits_total',
        help: 'Total search requests served from the Redis result cache',
        registers: [register]
    });
    const searchStage = new client.Histogram({
        name: 'search_stage_seconds',
        help: 'Duration of search pipeline stages in seconds',
        labelNames: ['stage'],
        buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5],
        registers: [register]
    });

    fastify.decorate('metrics', {
        register,
        searchRequests,
        searchCacheHits,
        searchStage
    });

    // Per-request timing. routeOptions.url keeps cardinality bounded (matched
    // pattern, not the concrete path).
    fastify.addHook('onRequest', async (request) => {
        request.metricsStart = process.hrtime.bigint();
    });

    fastify.addHook('onResponse', async (request, reply) => {
        if (!request.metricsStart) return;
        const seconds = Number(process.hrtime.bigint() - request.metricsStart) / 1e9;
        const route = request.routeOptions?.url || reply.request?.routerPath || request.url || 'unknown';
        httpHistogram
            .labels(request.method, route, String(reply.statusCode))
            .observe(seconds);
    });

    // Dedicated internal metrics server (kept off the public app port).
    const metricsServer = http.createServer(async (req, res) => {
        if (req.url === '/metrics') {
            try {
                const body = await register.metrics();
                res.writeHead(200, { 'Content-Type': register.contentType });
                res.end(body);
            } catch (err) {
                res.writeHead(500);
                res.end(String(err));
            }
            return;
        }
        if (req.url === '/healthz') {
            res.writeHead(200, { 'Content-Type': 'text/plain' });
            res.end('ok');
            return;
        }
        res.writeHead(404);
        res.end();
    });

    const port = resolveMetricsPort();
    metricsServer.listen(port, METRICS_HOST, () => {
        fastify.log.info(`Metrics exposed on http://${METRICS_HOST}:${port}/metrics`);
    });
    metricsServer.on('error', (err) => {
        fastify.log.error({ err }, 'Metrics server error');
    });

    fastify.addHook('onClose', async () => {
        await new Promise((resolve) => metricsServer.close(resolve));
    });
}

export default fp(metricsPlugin, { name: 'metrics' });

File: services/embedding/src/embedding/metrics.py
"""
Prometheus metrics for the embedding service.

Exposes RED metrics (Rate / Errors / Duration) under the SAME metric name and
label scheme as the Node services (`http_request_duration_seconds` with
method / route / status_code) so dashboards and alert rules stay uniform, plus
embedding-specific domain metrics.

Metrics are served on the app's own port at GET /metrics. nginx blocks the
public `/embed/metrics` path; Prometheus scrapes `embedding:8000/metrics`
directly on the internal Docker network.
"""

import time

from prometheus_client import (
    CONTENT_TYPE_LATEST,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)
from starlette.requests import Request
from starlette.responses import Response

# ── HTTP RED metrics ─────────────────────────────────────────────────
HTTP_REQUEST_DURATION = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    labelnames=("method", "route", "status_code"),
    buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 30),
)

# ── Embedding domain metrics ─────────────────────────────────────────
EMBEDDING_INFERENCE_SECONDS = Histogram(
    "embedding_inference_seconds",
    "Time spent generating embeddings for one request (seconds)",
    buckets=(0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 30, 60),
)

EMBEDDING_BATCH_SIZE = Histogram(
    "embedding_batch_size",
    "Number of texts per embedding request",
    buckets=(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 600),
)

EMBEDDING_IN_FLIGHT = Gauge(
    "embedding_in_flight_requests",
    "Embedding requests currently being processed",
)

EMBEDDING_REQUESTS_TOTAL = Counter(
    "embedding_requests_total",
    "Total embedding requests by mode and outcome",
    labelnames=("mode", "outcome"),  # mode: standalone|gateway ; outcome: success|error
)


def setup_metrics(app) -> None:
    """Attach the timing middleware and the /metrics endpoint to a FastAPI app."""

    @app.middleware("http")
    async def _track_requests(request: Request, call_next):
        # Skip self-scrape so /metrics does not inflate its own series.
        if request.url.path == "/metrics":
            return await call_next(request)

        start = time.perf_counter()
        status_code = 500
        try:
            response = await call_next(request)
            status_code = response.status_code
            return response
        finally:
            elapsed = time.perf_counter() - start
            route = request.scope.get("route")
            route_label = getattr(route, "path", None) or request.url.path
            HTTP_REQUEST_DURATION.labels(
                request.method, route_label, str(status_code)
            ).observe(elapsed)

    @app.get("/metrics", include_in_schema=False)
    async def metrics_endpoint() -> Response:
        return Response(content=generate_latest(), media_type=CONTENT_TYPE_LATEST)
